Hosting configuration: poll Atomic transfer status when loading the page (v2)

mpkelly opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Closes https://github.com/Automattic/wp-calypso/issues/72371
Related https://github.com/Automattic/wp-calypso/pull/77112 (I ported this original PR to the latest trunk and made some small improvements along the way)

## Proposed Changes

* Make sure the `/hosting-config` UI updates correctly when the Atomic transfer completes ( see https://github.com/Automattic/wp-calypso/issues/72371)
* When transferring, make the client-side status polling mechanism robust enough to survive a hard page refresh. 
* Now that the polling mechanism can pick up existing jobs, there does not seem to be a good reason to lock them into this page until the job is complete. I have removed the "click outside" logic in this PR. We still might want to include this. I would like @aidvu's input here, as they added the original restriction. 


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

There's been quite a few changes to boolean logic so there are two main things to test:

1. The state of `/hosting-config` for various plan + site type combos
2. The robustness of the Atomic transfer mechanism. See https://github.com/Automattic/wp-calypso/issues/72371 for inspiration. In short, it should be fine to go to another page and come back or do a hard refresh. The process should always complete and not get stuck. 

First, try this feature on prod and see that the process is broken and gets stuck. In my tests, it completed occasionally but mostly did not until I refreshed the page.  

Next, test viewing `/hosting-config` when you have:
 
**A Free site**: 
       * should see the upsell
       * should not see the activation banner
       * features should be masked

![image](https://github.com/Automattic/wp-calypso/assets/6851384/636bc92d-7375-485a-a8a7-2e068253baaf)
    
**a Creator plan but haven't gone Atomic**
       * You should see an activation banner but no upsell
       * Feature cards should be masked and unusable

![image](https://github.com/Automattic/wp-calypso/assets/6851384/1a7104f2-f745-499b-9a83-34c600b29878)
 
**an Atomic site**:
        * upsell should be gone
        * The activation banner should be gone 
        * All features should be unmasked and usable
        
<img width="1145" alt="Screenshot 2567-01-05 at 13 56 12" src="https://github.com/Automattic/wp-calypso/assets/6851384/a9e60841-87c5-4230-90e7-fb4b289a81db">
      
* **Test upgrading a Free plan to Creator**
       * should initially be like **A Free site** above
       * Then, when the process completes, it should be like **an Atomic site** above

* General robustness. We are polling the server for updates and should handle a refresh. Do your best to try and break it. 

[screen-capture (9).webm](https://github.com/Automattic/wp-calypso/assets/6851384/3c4e1a89-680f-404d-8f09-bf18021ccab0)


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

mpkelly committed:
Hosting config: check domain name to decide if activated

mpkelly committed:
Switch to the functional component + hook approach as other solution was not robust enough.

mpkelly committed:
- Don't keep polling if have an end state
- Check data.failureReason instead of data.code because latter doesn't get returned via React Query
- Disable showing of banner when no transferStatus available or no ftp-feature
- Get upsell banner working again
- Remove clickOutside which shouldn't be necessary (will discuss on GH)

zaguiini commented on client/my-sites/hosting/main.js:
```diff
@@ -184,207 +183,174 @@ const SidebarCards = ( { isBasicHostingDisabled } ) => {
+	const isSiteAtomic =
+		transferStatus === transferStates.COMPLETE || transferStatus === transferStates.COMPLETED;
+
+	if ( endStates.includes( transferStatus ) && transferRefetchEnabled ) {
+		// the result won't chnage so no need to call the API
+		setTransferRefetchEnabled( false );
 	}
```
Do we need a state here? I believe we can make a simple boolean variable instead. We can potentially inline this inside the `enabled` prop in `useAtomicTransferQuery` usage:

```ts
const transferRefetchEnabled = ! hasSftpFeature && ! endStates.includes( transferStatus )
```

mpkelly replied:
Thanks for suggesting this. We can't use `transferStatus` here because it comes from `useAtomicTransferQuery`, which also receives the `transferRefetchEnabled` during initialisation. If I switch to the code below (which I think is what you mean), I get a `ReferenceError: Cannot access 'transferStatus' before initialization`. 

But anyway, I think we should not disable refetching, so we can support the other scenarios you tested, like tabs that were opened before activation. 

```ts
	const { isTransferring, transferStatus } = useAtomicTransferQuery( siteSlug, {
		refetchInterval: 5000,
		enabled: endStates.includes( transferStatus ) && hasSftpFeature,
	} );
```

zaguiini commented on client/my-sites/hosting/main.js:
```diff
@@ -184,207 +183,174 @@ const SidebarCards = ( { isBasicHostingDisabled } ) => {
+				icon="bug"
+			/>
+		);
+		if ( ! isSiteAtomic && hasSftpFeature && ! isTransferring && transferStatus !== undefined ) {
```
It feels weird to have `! isTransferring && transferStatus !== undefined`. Can we consolidate these two variables, or use only one of them?

mpkelly replied:
Having `! isTransferring` was unnecessary, so I have removed that.

zaguiini commented on client/my-sites/hosting/main.js:
```diff
@@ -184,207 +183,174 @@ const SidebarCards = ( { isBasicHostingDisabled } ) => {
+	const banner = shouldShowUpgradeBanner ? getUpgradeBanner() : getAtomicActivationNotice();
 
-		const isBusinessTrial = isMigrationTrial || isHostingTrial;
+	const isBusinessTrial = isMigrationTrial || isHostingTrial;
```
I think we can use the selector here, just to avoid having duplicate logic around.

mpkelly replied:
I introduced `isSiteOnBusinessTrial` to simplfy things. I kept "Business" as it could add a lot of renaming noise and a revert seems likely based on current name-change performance.

zaguiini commented on client/state/automated-transfer/constants.ts:
```diff
@@ -15,6 +15,7 @@ export const transferStates = {
 	BACKFILLING: 'backfilling',
 	RELOCATING: 'relocating_switcheroo',
 	COMPLETE: 'complete',
+	COMPLETED: 'completed', // there seems to be two spellings for this state
```
🤦

mpkelly replied:
Right. I ported this from your original PR and have not fully investigated it yet. When a site has SSH enabled, the API returns `complete`, which was the original enum. Maybe we don't need this. I will try to answer that before we merge this PR.

zaguiini reviewed (commented):
I tested on three different tabs:

### The first one kicks off the transfer
The tab that kicked off the transfer **did not** update to reflect the successful transfer to Atomic :x:

### The second one was open before the transfer was initiated

The tab that was opened before the transfer started **did not**:
- update to reflect the transfer in progress state :x:
- update to reflected the transfer successful state :x:

### The third one was opened after the transfer was initiated

The tab that was open _while_ the transfer was in progress **did update** to reflect the successful transfer to Atomic ✅ 

---

The video below demonstrates the testing. Another thing from the video is that after I kicked off the transfer, I had to wait five seconds until I saw the transferring notice. I guess that's because of the polling interval, but could we refetch automatically after starting to transfer?

https://github.com/Automattic/wp-calypso/assets/26530524/eb826f32-d3fb-4d6e-bac2-1a91c0487b22

danielbachhuber commented:
Without looking at the code too deeply, would it make sense to reconcile this implementation with https://github.com/Automattic/wp-calypso/pull/83545?

mpkelly commented:
@zaguiini, thanks for the great review. There is some redundancy in the logic that I will tidy up. Some renaming or aliasing will also make the code easier to read. 

>The tab that kicked off the transfer did not update to reflect the successful transfer to Atomic ❌

This surprised me because I worked on this across a couple of days and never encountered any issues - the process was always completed. I used a single Chrome tab; otherwise, it was the same as the video. I will take another look today. 

> The second one was open before the transfer was initiated

I didn't plan to support this and even made the polling mechanism short-circuit if the first response was an end state. It's easy to support if we're cool with a polling mechanism that keeps running. 

I will address some of your comments and tag you when ready for a re-review.

mpkelly committed:
1. Remove refetch optimisation so previously updated tabs update
2. Rename some variables to make logic easier to understand
3. Stop Activate banner blink when polling is in progress by adding hasFetchedTransferStatusAtLeastOnce

mpkelly committed:
Add isSiteOnBusinessTrial to tidy up hosting/migration trial logic.

mpkelly commented:
[screen-capture (11).webm](https://github.com/Automattic/wp-calypso/assets/6851384/6c897446-c447-4805-8575-ee5f14e59edc)

@zaguiini and anyone else using a browser other than Chrome, please retest. The test scenarios https://github.com/Automattic/wp-calypso/pull/85937#pullrequestreview-1806313609 should all be working now. If you check the video at around 36 seconds, you will see the banner change briefly. I have found it hard to fix this issue, and I am eager to get this out as it's quite broken in prod. Maybe we can fix this in a follow-up PR?

mpkelly committed:
Alias logic into something with a nicer name

zaguiini commented:
>It's easy to support if we're cool with a polling mechanism that keeps running.

I don't think it'd hurt. Hard to imagine someone keeping that tab open forever, but even harder to think of someone that will have two HC tabs open in pre-transfer state 🤔

zaguiini commented on client/state/atomic-transfer/use-atomic-transfer-query.ts:
```diff
@@ -0,0 +1,70 @@
+	return [ 'sites', siteSlug, 'atomic', 'transfers', 'latest' ];
+}
+
+export const useAtomicTransferQuery = (
```
I wonder if we need to define this hook since several mechanisms already exist:
- [`useAtomicTransfer`](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/checkout-thank-you/marketplace/use-atomic-transfer.tsx#L16) hook
- [`WithAtomicTransfer` HOC](https://github.com/Automattic/wp-calypso/blob/trunk/client/sites-dashboard/components/with-atomic-transfer.tsx#L40) (has a different hook inside)
- [an automated transfer handler](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/data-layer/wpcom/sites/automated-transfer/status/index.js#L47) (this is what `hosting/main.js` is using currently)

Seems like we could also unify these solutions.

mpkelly replied:
We can surely tidy these up. I think `use-atomic-transfer.tsx` could be simplified by using the same hook we're adding here.

I think unification can be done as a separate PR as it will only delay this longer. Daniel posted another PR we can consider for unification, too.

zaguiini replied:
We could revisit this as a follow-up.

zaguiini commented on client/my-sites/hosting/main.js:
```diff
@@ -184,207 +176,173 @@ const SidebarCards = ( { isBasicHostingDisabled } ) => {
-					</>
-				);
-			}
+	const isSiteAtomic =
+		transferStatus === transferStates.COMPLETE || transferStatus === transferStates.COMPLETED;
```
Does that necessarily mean the site is Atomic? There are several ways to check if the site is Atomic, not necessarily relying on the transfer status:
- `is_automated_transfer` site option
- `is_wpcom_atomic` site option (seems to be the most reliable and tied to WPCOM)

mpkelly replied:
The `complete[d]` state does mean the site is Atomic. If you revert the site, that also uses a transfer, which changes the status to `reverted`.  There will be no transfer if the site is Simple and never went Atomic. If an error happens, the status will be `error`. There are some other states in `transferStates` to cover other scenarios. 

There are other flags that require a different query, whereas we get the final status + interim status updates via this one query.

zaguiini replied:
Ok. I'm still skeptical about data derivation (Atomic meaning the site has transfer state), but it worked in testing, so why not try it as the feature was broken for a while.

I'm considering a case where the transfer was made a while ago and the transfer record has been cleaned up. In this case there won't be a transfer. So WDYT of adding `|| site.options.is_wpcom_atomic` as a final check at the end, just to be sure?

zaguiini commented on client/state/atomic-transfer/use-atomic-transfer-query.ts:
```diff
@@ -0,0 +1,70 @@
+		// Will happen for new sites when no transfer exists: a 404 is returned
+		// by the API which becomes failureReason in React Query
+		return {
+			transferStatus: transferStates.NONE,
```
I'm assuming the request returns a successful status even if there was an error during the transfer. Can we double-check it?

mpkelly replied:
The HTTP response code is `200`. However, the payload looks like this:

```json
{
    "body": {
        "code": "no_transfer_record",
        "message": "Transfer record not found for blog `227821875`",
        "data": {
            "status": 404
        }
    },
    "status": 404,
    "headers": {
        "Allow": "GET"
    }
}
```
And React Query parses it to:

```js
{
  "status": "loading",
  "fetchStatus": "fetching",
  "isLoading": true,
  "isSuccess": false,
  "isError": false,
  "isInitialLoading": true,
  "dataUpdatedAt": 0,
  "error": null,
  "errorUpdatedAt": 0,
  "failureCount": 3,
  "failureReason": {
    "path": "/sites/mk9287907.wordpress.com/atomic/transfers/latest",
    "apiNamespace": "wpcom/v2",
    "method": "GET",
    "apiVersion": "1.1",
    "query": "",
    "callback": "207c6f39-c612-448f-acde-7684177e1f48",
    "supports_args": true,
    "supports_error_obj": true,
    "supports_progress": true,
    "name": "NotFoundError",
    "statusCode": 404,
    "status": 404,
    "message": "Transfer record not found for blog `227821875`",
    "code": "no_transfer_record",
    "data": {
      "status": 404
    }
  },
  "errorUpdateCount": 0,
  "isFetched": false,
  "isFetchedAfterMount": false,
  "isFetching": true,
  "isRefetching": false,
  "isLoadingError": false,
  "isPaused": false,
  "isPlaceholderData": false,
  "isPreviousData": false,
  "isRefetchError": false,
  "isStale": true
}
```

zaguiini replied:
We could revisit this as a follow-up.

zaguiini reviewed (commented):
All three scenarios described in my previous testing worked this time 👍

I didn't manage to reproduce the flick you reported, weird.

mpkelly commented:
FYI, I am going to get these tests passed and add any more I think we need. 

<img width="1170" alt="Screenshot 2567-01-09 at 11 48 40" src="https://github.com/Automattic/wp-calypso/assets/6851384/44928f6a-0ed9-49b3-8663-3117f040e6ab">

zaguiini reviewed (commented):
I like the progress of this PR. Once the tests are here and we can verify that it will work forever with no regressions, we can go ahead and merge this.

mpkelly committed:
Fix hosting tests by mocking API to return transfer status

mpkelly committed:
Fix isActive by adding inquiring and conflict states to end state array. I am not sure if these are ever returned by the server but removing them breaks the hosting-config/activate UI, so just going to keep them.

mpkelly commented:
I think tests are still failing because of this. I will fix it tomorrow. 

```
 console.error
      Warning: An update to InlineSupportLink inside a test was not wrapped in act(...).
      
      When testing, code that causes React state updates should be wrapped into act(...):
      
      act(() => {
        /* fire events that update state */
      });
      /* assert on the output */
```

zaguiini reviewed (approved):
Looks good to me. Once the test passes, feel free to merge 👍

mpkelly committed:
Fix leaky async resource issue blocking Calypso build by waiting for async jobs to complete.

mpkelly committed:
Unmock after test not before.

mpkelly committed:
Temp change: ignore hosting tests to debug failing build (it's not clear which file is causing the issue)

mpkelly committed:
Revert debugging change. The problem is not in the test file am I mostly changing so let's try a rebase next.

mpkelly commented:
The build is still failing. 🤔 I disabled the main test file I am working with, and the build still fails. So, I'm hoping this rebase will fix it. 

<img width="1201" alt="Screenshot 2567-01-10 at 11 34 36" src="https://github.com/Automattic/wp-calypso/assets/6851384/410586ae-655c-4ef6-be03-e5aa8b228578">

>   A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.

mpkelly committed:
Fix the error. The real one...
